README.md
# pyghotess
Python / Ghostscript / Tesseract fast image PDF OCR processing

## Description
This component allows more fine-tuning, control and parallelisation of Tesseract OCR processing then other available solutions (apache Tika for example).
It is only applicable to PDF files though.

### Advantages
- Configure the parameters to the task at hand : is the original material of good quality ? Is the orientation bad ? The resolution ?
- Distribute the OCR processing over multiple CPU's

### Disadvantage
- Less flexible: only works with PDF files

### Other ideas
- Streaming of output : get first pages while other pages are still processed
- Per-page configuration

## Installation
From the root directory of in your local copy (after cloning the repository):
```bash
# poetry
> poetry install
```
### Poetry installation
Poetry installs a local, isolated python environment, to avoid conflicts with your system's python modules. See [here](https://python-poetry.org/docs/).

A recent [python](https://www.python.org/downloads/) version (>=3.7) will also be needed.

## Usage
Can be used with **poetry**, **docker** CLI or as **API** (Poetry or Docker).

### CLI usage
```bash
Usage: run.py [OPTIONS] COMMAND [ARGS]...

Options:
  -d, --debug TEXT  Activate debugging flag
  --help            Show this message and exit.

Commands:
  extract  EXTRACT Extract PDF files to PNG images
      -f, --file TEXT     Input image-PDF file
      -o, --out-dir TEXT  Root directory for temp directory

  ocr      OCR Parse directory with PNG images by OCR processing and output...
      -d, --dir TEXT  Directory with image files

  process  PARSE Parse Image PDF file, output resulting text.
      -f, --file TEXT  Input image-PDF file

# Example (with poetry)
$ poetry run ./run.py process -f test2.pdf
```

### API Usage
By default, the API binds to port 5501 (settings are set by environment variables or by using a config file, see `config_default.toml`)
```bash
# Run locally in debug mode
> poetry run api --debug
# Run locally with config file
> poetry run api --config config.toml
```
## Contributing
Pull requests are welcome. For major changes, please open an issue first to discuss what you would like to change.

Please make sure to update tests as appropriate.

## Updating docker
The `docker_tool.sh` script allows for reasy updating of docker hub images. Update with your own repo settings if needed.

```
# Build
$ ./docker_tool.sh build [VERSION]
# Test
$ ./docker_tool.sh test [VERSION]
# Publish
$ ./docker_tool.sh publish [VERSION]
# Tag latest version
$ ./docker_tool.sh latest [VERSION]
```

### Todo:
- Finish streaming OCR results with the websocket
- Shorten TTFR (Time To First Result) even more

### Testing
Pyghotess uses behave BDD testing. Features and steps are specified in the `tests/behave` folder and can be run this way:
```bash
$ poetry run behave tests/behave
```


CHANGELOG.md
CHANGELOG
=========

This changelog references the relevant changes done in this project.

This project adheres to [Semantic Versioning](http://semver.org/) 
and to the [CHANGELOG recommendations](http://keepachangelog.com/).

## Alpha Development

### [0.2.0] - (2020-12-29)
- Added changelog
- Adding websocket support for streaming tesseract output in real time (page by page)


tests/behave/fixtures.py
import logging
from time import sleep
from multiprocessing import Process
import uvicorn
from behave import fixture
from pyghotess import api

PORT = 5555
HOST = '127.0.0.1'


def run_server():
    uvicorn.run(api, port=PORT, host=HOST)


@fixture
def server_api(_context, **_kwargs):
    proc = Process(target=run_server, args=(), daemon=True)
    logging.info('Starting server subprocess')
    proc.start()
    sleep(1.0)
    yield
    logging.info('Stopping server subprocess')
    proc.kill()






tests/behave/environment.py
# pylint: skip-file
from behave import use_fixture
from fixtures import server_api, PORT, HOST


# Fixture Settings
# ##################################################################
def before_tag(context, tag):
    if tag == 'fixture.api':
        use_fixture(server_api, context)
    pass


# Setup & Breakdown
# ##################################################################
def before_all(context):
    context.api_port = PORT
    context.api_host = HOST
    pass


def after_all(context):
    pass


def before_scenario(context, scenario):
    pass


def after_scenario(context, scenario):
    pass


tests/behave/steps/ws_steps.py
from behave import when, given, then
import json
import os
import websocket


@given(u'a connection to the websocket')
def step_impl(context):
    context.ws = websocket.WebSocket()
    context.ws.connect(f'ws://{context.api_host}:{context.api_port}/ws')
    assert context.ws


@when(u'I send a simple test message')
def step_impl(context):
    context.ws.send(json.dumps({
        'payload': 'coucou',
        'action': 'test',
    }))


@then(u'I receive a response')
def step_impl(context):
    result = json.loads(context.ws.recv())
    assert result['action'] == 'test'
    assert result['payload'] == 'coucou'


@when(u'I close the connection')
def step_impl(context):
    context.ws.close()


@then(u'the connection is closed')
def step_impl(context):
    assert context.ws.connected is False


@when(u'I send a test file')
def step_impl(context):
    fpath = os.path.join(os.getcwd(), 'resources/test_fr_7_img.pdf')
    context.ws.send(json.dumps({
        'action': 'upload',
        'filename': 'test_fr_7_img.pdf',
    }))
    result = json.loads(context.ws.recv())
    assert result['action'] == 'upload'

    i = 1
    with open(fpath, 'rb') as f:
        while chunk := f.read(1024 * 1024):
            print(f'sending chunk {i}')
            context.ws.send_binary(chunk)
            i += 1
        context.ws.send_binary('')


@then(u'I can query the file status')
def step_impl(context):
    context.ws.send(json.dumps({'action': 'get_file_status'}))
    result = json.loads(context.ws.recv())
    print(result)
    assert result['action'] == 'get_file_status'
    assert result['payload'] == 'uploaded'


@when(u'I wait for the results')
def step_impl(context):
    context.received = 0
    context.pages = []
    while True:
        result = json.loads(context.ws.recv())
        if result['action'] == 'done':
            break
        if result['action'] == 'get_file_status':
            print(f"status is {result['payload']}")
            continue
        context.received += 1
        assert result['action'] == 'page_extract'
        assert result['meta']['page']
        print(f"received page {result['meta']['page']} (received {context.received})")
        context.pages.append(result['payload'])
        context.ws.send(json.dumps({'action': 'get_file_status'}))


@then(u'I receive all 7 pages')
def step_impl(context):
    print(context.received)
    print(context.pages)
    assert context.received == 7


run.py
#!/usr/bin/env python3
#  Pyghotess, fast image-PDF OCR Processing
#     Copyright (C) 2020    Pieterjan Montens
#     This program is free software: you can redistribute it and/or modify
#     it under the terms of the GNU General Public License as published by
#     the Free Software Foundation, either version 3 of the License, or
#     (at your option) any later version.
#     This program is distributed in the hope that it will be useful,
#     but WITHOUT ANY WARRANTY; without even the implied warranty of
#     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#     GNU General Public License for more details.
#     You should have received a copy of the GNU General Public License
#     along with this program.  If not, see <https://www.gnu.org/licenses/>.
#

import click
import sys
import logging
import os
import time
import math
import asyncio
import pyghotess as pgt

logger = logging.getLogger(__name__)
logger.setLevel(logging.getLevelName('INFO'))
logger.addHandler(logging.StreamHandler())

config = {
    'extract': {
        'resolution': os.getenv('RESOLUTION', 200),
        'alphaBits': 4,
    },
    'ocr': {
        'resolution': os.getenv('RESOLUTION', 200),
        'workers': os.getenv('WORKERS', 2),
        'psm': 4
    }
}


@click.group()
@click.option('--debug', '-d', 'debug', default=False,
              help="Activate debugging flag")
def run(debug):
    if debug:
        logger.setLevel(logging.getLevelName('DEBUG'))
        logger.info('Debugging Enabled')


@run.command()
@click.option('--file', '-f', 'fileIn', default="",
              help="Input image-PDF file")
@click.option('--out-dir', '-o', 'outDir', default="",
              help="Root directory for temp directory")
def extract(fileIn, outDir):
    """
    EXTRACT
    Extract PDF files to PNG images
    """
    dirPath = os.path.dirname(os.path.realpath(__file__))
    filePath = os.path.join(dirPath, fileIn)
    if outDir == "":
        outDir = os.getcwd()

    with pgt.pdf2png(config, filePath, outDir, cleanUp=False) as (taskid, path):
        logger.info('Extracted PDF file to directory:/n %s', path)
        sys.stdout.write(path)


@run.command()
@click.option('--dir', '-d', 'directory', default="",
              help="Directory with image files")
def ocr(directory):
    """
    OCR
    Parse directory with PNG images by OCR processing
    and output resulting text.
    """
    dir_path = os.path.dirname(os.path.realpath(__file__))
    work_path = os.path.join(dir_path, directory)

    payload = []
    for fname in os.listdir(directory):
        order = int(fname[3:5])
        fpath = os.path.join(work_path, fname)
        payload.append((order, fpath))

    loop = asyncio.get_event_loop()
    [result] = loop.run_until_complete(asyncio.gather(pgt.process(payload, config)))
    sys.stdout.write(pgt.outputParse(result))


@run.command()
@click.option('--file', '-f', 'fileIn', default="",
              help="Input image-PDF file")
def process(fileIn):
    """
    PARSE
    Parse Image PDF file, output resulting text.
    (combines Extract and Ocr commands)
    """

    dirPath = os.path.dirname(os.path.realpath(__file__))
    filePath = os.path.join(dirPath, fileIn)
    outDir = os.getcwd()

    with pgt.pdf2png(config, filePath, outDir) as (taskid, path):

        payload = []
        for fname in os.listdir(path):
            order = int(fname[3:5])
            fpath = os.path.join(path, fname)
            payload.append((order, fpath))

        loop = asyncio.get_event_loop()
        [result] = loop.run_until_complete(asyncio.gather(pgt.process(payload, config)))

    sys.stdout.write(pgt.outputParse(result))


@run.command()
def benchmark():
    """
    Run a quick platform benchmark
    """
    dirPath = os.path.dirname(os.path.realpath(__file__))
    filePath = os.path.join(dirPath, './resources/test_fr_7_img.pdf')
    outDir = os.getcwd()
    Score = lambda x: math.floor(7 / (time.time() - x) * 100)

    starttime = time.time()
    config['ocr']['workers'] = 1
    with pgt.pdf2png(config, filePath, outDir) as (taskid, path):
        payload = []
        for fname in os.listdir(path):
            order = int(fname[3:5])
            fpath = os.path.join(path, fname)
            payload.append((order, fpath))

        loop = asyncio.get_event_loop()
        [result] = loop.run_until_complete(asyncio.gather(pgt.process(payload, config)))

    worker1_time = Score(starttime)

    starttime = time.time()
    config['ocr']['workers'] = 2
    with pgt.pdf2png(config, filePath, outDir) as (taskid, path):
        payload = []
        for fname in os.listdir(path):
            order = int(fname[3:5])
            fpath = os.path.join(path, fname)
            payload.append((order, fpath))

        loop = asyncio.get_event_loop()
        [result] = loop.run_until_complete(asyncio.gather(pgt.process(payload, config)))

    worker2_time = Score(starttime)

    starttime = time.time()
    config['ocr']['workers'] = 'auto'
    with pgt.pdf2png(config, filePath, outDir) as (taskid, path):
        payload = []
        for fname in os.listdir(path):
            order = int(fname[3:5])
            fpath = os.path.join(path, fname)
            payload.append((order, fpath))

        loop = asyncio.get_event_loop()
        [result] = loop.run_until_complete(asyncio.gather(pgt.process(payload, config)))

    workera_time = Score(starttime)

    sys.stdout.write(f"""
    1 Worker score : {worker1_time}
    2 Worker score : {worker2_time}
    Auto Worker score : {workera_time}

    Final: {int((worker1_time + worker2_time + workera_time)/3)}
    """)
    # print(worker3_time)


if __name__ == "__main__":
    run()


docker_tool.sh
#!/usr/bin/env bash
while getopts ":h" opt; do
  case ${opt} in
    h )
      echo "Usage:"
      echo "    docker-tool.sh -h        Display this help message."
      echo "    docker-tool.sh build <VERSION>    Builds image"
      echo "    docker-tool.sh test <VERSION>    Tests image"
      echo "    docker-tool.sh publish <VERSION>  Publishes image to Docker Hub"
      echo "    docker-tool.sh latest <VERSION>  Tags images as latest on Docker Hub."
      echo "    docker-tool.sh run <VERSION>  Run the image locally"
      exit 0
      ;;
   \? )
     echo "Invalid Option: -$OPTARG" 1>&2
     exit 1
     ;;
  esac
done


test_docker_image() {
     docker run -d --name "$1" -p 5005:6006 -e HOST='0.0.0.0' berzemus/pyghotess:"$1"
     sleep 2
     url=http://localhost:5005/
     status=$(curl --get --location --connect-timeout 5 --write-out %{http_code} --silent --output /dev/null ${url})

     if [[ $status == '200' ]]
     then
      echo "$(tput setaf 2)Image: berzemus/pyghotess:${1} - Passed$(tput sgr0)"
      docker kill "$1"
      docker rm "$1"
     else
      echo "$(tput setaf 1)Image: berzemus/pyghotess:${1} - Failed$(tput sgr0)"
      docker kill "$1"
      docker rm "$1"
      exit 1
     fi
}

shift $((OPTIND -1))
subcommand=$1; shift
version=$1; shift

case "$subcommand" in
  build)
    # Build slim version with minimal dependencies
    docker build -t berzemus/pyghotess:${version}-api  -f ./Docker/Dockerfile_API --no-cache .
    #docker build -t berzemus/pyghotess:${version}-cli  -f ./Docker/Dockerfile_CLI --no-cache .
    ;;

  test)
    # Test the images
    test_docker_image ${version}-api
    # test_docker_image "${version}-cli"
    ;;

  publish)
    # Push the build images
    docker push berzemus/pyghotess:${version}-api
    # docker push berzemus/pyghotess:${version}-cli
    ;;

  latest)
    # Update the latest tags to point to supplied version
    docker tag berzemus/pyghotess:${version}-api berzemus/pyghotess:latest-api
    docker push berzemus/pyghotess:latest-api
    # docker tag berzemus/pyghotess:${version}-cli berzemus/pyghotess:latest-ocr
    # docker push berzemus/pyghotess:latest-ocr
    ;;

  run)
    # Run the image
    docker run -it --rm -p 6006:6006 berzemus/pyghotess:${version}-api
    ;;

esac


pyghotess/lib_misc.py
#  Pyghotess, fast image-PDF OCR Processing
#     Copyright (C) 2020    Pieterjan Montens
#     This program is free software: you can redistribute it and/or modify
#     it under the terms of the GNU General Public License as published by
#     the Free Software Foundation, either version 3 of the License, or
#     (at your option) any later version.
#     This program is distributed in the hope that it will be useful,
#     but WITHOUT ANY WARRANTY; without even the implied warranty of
#     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#     GNU General Public License for more details.
#     You should have received a copy of the GNU General Public License
#     along with this program.  If not, see <https://www.gnu.org/licenses/>.
#
from datetime import datetime
import pytz
import math


def starttime_get():
    return datetime.now(pytz.utc)


def status_get(start_time, version, counter):
    now = datetime.now(pytz.utc)
    delta = now - start_time
    delta_s = math.floor(delta.total_seconds())
    return {
        'all_systems': 'nominal',
        'id': __name__,
        'timestamp': str(now),
        'online_since': str(start_time),
        'online_for_seconds': delta_s,
        'api_version': version,
        'api_counter': counter,
    }


pyghotess/lib_parse.py
#  Pyghotess, fast image-PDF OCR Processing
#     Copyright (C) 2020    Pieterjan Montens
#     This program is free software: you can redistribute it and/or modify
#     it under the terms of the GNU General Public License as published by
#     the Free Software Foundation, either version 3 of the License, or
#     (at your option) any later version.
#     This program is distributed in the hope that it will be useful,
#     but WITHOUT ANY WARRANTY; without even the implied warranty of
#     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#     GNU General Public License for more details.
#     You should have received a copy of the GNU General Public License
#     along with this program.  If not, see <https://www.gnu.org/licenses/>.
#
import os
import uuid
import asyncio
import logging
import shutil
import math
import subprocess
import multiprocessing
from contextlib import contextmanager

logger = logging.getLogger(__name__)
logger.setLevel(logging.getLevelName('INFO'))
logger.addHandler(logging.StreamHandler())


def outputParse(listIn):
    listIn.sort(key=lambda x: x[0])
    listOut = [x for _y, x in listIn]
    return '\n'.join(listOut)


@contextmanager
def pdf2png(config, filePath, outDir=False, *, cleanUp=True):
    """
    pdf2png
    - Make a temporary directory
    - Extract PDF pages to PNG with GhostScript
    - Yield directory path
    - Finally and in all cases : clean up temp directory
    """
    taskid = str(uuid.uuid4())

    if not outDir:
        outDir = os.getcwd()

    tempPath = os.path.join(outDir, taskid)
    os.mkdir(tempPath)

    cfg = config.get('extract', {'resolution': 200, 'alphaBits': 4})
    logger.info('pdf2png: %s', cfg)
    cmd = [
        'gs',
        '-q',
        '-sDEVICE=png16m',
        f'-dTextAlphaBits={cfg["alphaBits"]}',
        f'-o{tempPath}/gs-%02d.png',
        '-dNOPAUSE',
        f'-r{cfg["resolution"]}',
        filePath,
    ]
    subprocess.run(cmd)

    try:
        yield (taskid, tempPath)

    finally:
        if cleanUp:
            shutil.rmtree(tempPath)


async def worker(name, cfg, language, queue_in, queue_out):
    while True:
        try:
            order, fpath = await queue_in.get()
            logger.info('Worker %s starting on file %s', name, order)
            logger.debug(cfg)

            cmd = [
                'tesseract',
                '--psm', str(cfg['psm']),
                '-l', language,
                '--dpi', str(cfg['resolution']),
                '--oem', '1',
                fpath,
                'stdout',
            ]
            await asyncio.sleep(.1)

            proc = await asyncio.create_subprocess_shell(
                ' '.join(cmd),
                stdout=asyncio.subprocess.PIPE)
            # result = subprocess.run(cmd, stdout=subprocess.PIPE)
            stdout, _ = await proc.communicate()
            queue_out.put_nowait((order, stdout.decode()))
            queue_in.task_done()
        except Exception as e:
            logger.exception(e)


async def process(payloads, config={}, language='fra'):
    queue_in = asyncio.Queue()
    queue_out = asyncio.Queue()

    logger.debug('Filling queue in')
    cfg = config.get('ocr', {'psm': 4, 'dpi': 200, 'workers': 2})

    logger.info('process: %s', cfg)
    for payload in payloads:
        queue_in.put_nowait(payload)

    logger.debug('Creating tasks')
    tasks = []
    if cfg['workers'] == 'auto':
        workers = int(math.ceil(multiprocessing.cpu_count() / 4))
    else:
        workers = int(cfg['workers'])

    for i in range(workers):
        task = asyncio.create_task(worker(
            f'w-{i}',
            cfg,
            language,
            queue_in,
            queue_out
        ))
        tasks.append(task)

    logger.debug('Waiting job execution')
    await queue_in.join()

    logger.debug('Ending workers')
    for task in tasks:
        task.cancel

    resp = []
    logger.debug('Getting results')
    try:
        while True:
            resp.append(queue_out.get_nowait())
    except asyncio.queues.QueueEmpty:
        pass

    # Cancel our worker tasks.
    logger.debug('Cleaning up')
    for task in tasks:
        task.cancel()
    # Wait until all worker tasks are cancelled.
    await asyncio.gather(*tasks, return_exceptions=True)

    return resp


async def processWithGatherer(payloads, config, gatherer, language='fra'):
    queue_in = asyncio.Queue()
    queue_out = asyncio.Queue()

    async def gth(queue_gath):
        while True:
            try:
                order, test = await queue_gath.get()
                logger.debug('Gatherer received %s', order)
                await gatherer(order, test)
                queue_gath.task_done()
            except Exception as e:
                logger.exception(e)

    logger.debug('Filling queue in')
    cfg = config.get('ocr', {'psm': 4, 'dpi': 200, 'workers': 2})

    logger.info('process: %s', cfg)
    for payload in payloads:
        queue_in.put_nowait(payload)

    logger.debug('Creating tasks')
    tasks = []
    if cfg['workers'] == 'auto':
        workers = int(math.ceil(multiprocessing.cpu_count() / 4))
    else:
        workers = int(cfg['workers'])

    for i in range(workers):
        task = asyncio.create_task(worker(
            f'w-{i}',
            cfg,
            language,
            queue_in,
            queue_out
        ))
        tasks.append(task)

    gath_job = asyncio.create_task(gth(queue_out))
    tasks.append(gath_job)

    logger.debug('Waiting job execution')
    await queue_in.join()
    logger.debug('Job done, ending workers')

    for task in tasks:
        task.cancel()
    # Wait until all worker tasks are cancelled.
    await asyncio.gather(*tasks, return_exceptions=True)


pyghotess/main.py
#!/usr/bin/env python3
#  Pyghotess, fast image-PDF OCR Processing
#     Copyright (C) 2020    Pieterjan Montens
#     This program is free software: you can redistribute it and/or modify
#     it under the terms of the GNU General Public License as published by
#     the Free Software Foundation, either version 3 of the License, or
#     (at your option) any later version.
#     This program is distributed in the hope that it will be useful,
#     but WITHOUT ANY WARRANTY; without even the implied warranty of
#     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#     GNU General Public License for more details.
#     You should have received a copy of the GNU General Public License
#     along with this program.  If not, see <https://www.gnu.org/licenses/>.
#
import logging
import shutil
import os
import argparse
import uvicorn
import toml
import tempfile
import json
from fastapi import (
    Depends,
    FastAPI,
    HTTPException,
    File,
    UploadFile,
    WebSocket,
    WebSocketDisconnect,
)
from typing import List
from fastapi.responses import PlainTextResponse
from starlette.middleware.cors import CORSMiddleware
import pyghotess.lib_misc as lm
import pyghotess.lib_parse as lp


# ################################################### SETUP AND ARGUMENT PARSING
# ##############################################################################
logger = logging.getLogger(__name__)
logger.setLevel(logging.getLevelName('INFO'))
logger.addHandler(logging.StreamHandler())
dir_path = os.path.dirname(os.path.realpath(__file__))

config = {
    'server': {
        'host': os.getenv('HOST', '127.0.0.1'),
        'port': int(os.getenv('PORT', '5501')),
        'log_level': os.getenv('LOG_LEVEL', 'info'),
        'timeout_keep_alive': 0,
    },
    'proxy_prefix': os.getenv('PROXY_PREFIX', ''),
    'log_level': 'info',
    'access_key': os.getenv('ACCESS_KEY', False),
    'extract': {
        'resolution': os.getenv('RESOLUTION', 200),
        'alphaBits': 4,
    },
    'ocr': {
        'resolution': os.getenv('RESOLUTION', 200),
        'workers': os.getenv('WORKERS', 2),
        'psm': 4
    }
}


COUNTER = 0
VERSION = 3
START_TIME = lm.starttime_get()

# ################################################################## APP CONFIG
# #############################################################################


app = FastAPI(root_path=config['proxy_prefix'])
app.add_middleware(
    CORSMiddleware,
    allow_origins=["*"],
    allow_credentials=True,
    allow_methods=["*"],
    allow_headers=["*"],
)


class ConnectionManager:
    def __init__(self):
        self.active_connections: List[WebSocket] = []

    async def connect(self, websocket: WebSocket):
        await websocket.accept()
        self.active_connections.append(websocket)

    def disconnect(self, websocket: WebSocket):
        self.active_connections.remove(websocket)

    async def send_personal_message(self, message: str, websocket: WebSocket):
        await websocket.send_text(message)

    async def broadcast(self, message: str):
        for connection in self.active_connections:
            await connection.send_text(message)


manager = ConnectionManager()


# ############################################################### SERVER ROUTES
# #############################################################################
@app.get("/")
def root():
    return lm.status_get(START_TIME, VERSION, COUNTER)


@app.post("/process")
async def process(rawFile: UploadFile = File(...)):
    """
    Process
    Execute OCR Processing on uploaded file
    """
    with tempfile.NamedTemporaryFile() as tFile:
        shutil.copyfileobj(rawFile.file, tFile.file)

        with lp.pdf2png(config, tFile.name) as (taskid, path):
            payload = []
            for fname in os.listdir(path):
                order = int(fname[3:5])
                fpath = os.path.join(path, fname)
                payload.append((order, fpath))

            result = await lp.process(payload, config)
            text = lp.outputParse(result)

    return PlainTextResponse(text)


@app.websocket("/ws")
async def websocket_endpoint(ws: WebSocket):
    status = {}
    await manager.connect(ws)
    try:
        while True:
            data = await ws.receive_json()
            if data['action'] == 'test':
                await ws.send_json(data)
            if data['action'] == 'get_file_status':
                await ws.send_json({
                    'action': 'get_file_status',
                    'payload': status['file']['status']
                })
            if data['action'] == 'upload':
                with tempfile.NamedTemporaryFile() as tFile:
                    status['file'] = {
                        'status': 'uploading',
                        'file': tFile
                    }
                    logger.debug('ready to receive bytes')
                    await ws.send_json({'action': 'upload'})
                    while True:
                        rawBytes = await ws.receive_bytes()
                        tFile.write(rawBytes)
                        if len(rawBytes) == 0:
                            logger.info("File received")
                            break
                        logger.debug("Wrote %s bytes", len(rawBytes))
                    tFile.seek(0)
                    status['file']['status'] = 'uploaded'

                    async def fun(order, text):
                        logger.debug(f'sending text {order} to ws')
                        await ws.send_json({
                            'action': 'page_extract',
                            'meta': {'page': order},
                            'payload': text,
                        })

                    with lp.pdf2png(config, tFile.name) as (taskid, path):
                        payload = []
                        for fname in os.listdir(path):
                            order = int(fname[3:5])
                            fpath = os.path.join(path, fname)
                            payload.append((order, fpath))
                        payload.sort(key=lambda x: x[0])

                        status['file']['status'] = 'OCR processing'
                        await lp.processWithGatherer(payload, config, fun)

                await ws.send_json({'action': 'done'})

    except WebSocketDisconnect:
        manager.disconnect(ws)
        logger.debug("Disconnecting")


# ##################################################################### STARTUP
# #############################################################################
def main():
    global config

    parser = argparse.ArgumentParser(description='Pyghotess API server process')
    parser.add_argument('--config', dest='config', help='config file', default=None)
    parser.add_argument('--debug', dest='debug', action='store_true', default=False, help='Debug mode')
    args = parser.parse_args()

    if args.config:
        t_config = toml.load(['config_default.toml', args.config])
    else:
        t_config = toml.load('config_default.toml')

    config = {**config, **t_config}

    if args.debug:
        logger.setLevel(logging.getLevelName('DEBUG'))
        logger.debug('Debug activated')
        config['log_level'] = 'debug'
        config['server']['log_level'] = 'debug'
        logger.debug('Arguments: %s', args)
        logger.debug('config: %s', toml.dumps(config))

    uvicorn.run(
        app,
        **config['server']
    )


if __name__ == "__main__":
    main()
